proposals/metadata.md

```diff
@@ -0,0 +1,103 @@
+<!--
+Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License"). You may
+not use this file except in compliance with the License. A copy of the
+License is located at
+
+     http://aws.amazon.com/apache2.0/
+
+or in the "license" file accompanying this file. This file is distributed
+on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+express or implied. See the License for the specific language governing
+permissions and limitations under the License.
+-->
+
+### Container metadata feature
+In order to solve the issues related to container metadata like: [#288](https://github.com/aws/amazon-ecs-agent/issues/288) and [#456](https://github.com/aws/amazon-ecs-agent/issues/456), we are considering adding a new feature "Container Metadata" similar to [EC2 Instance Metadata Service](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) in ECS agent to provide the container metadata.
+
+By adding this feature, you can query the information about the task, container and container instance from within the container or from the container instance by reading the metadata file for each container. The following information will be available from this feature:
+ * Information from Docker:
+  * Container Network Information
+  * Container ID from Docker
+  * Container Name
+  * Image Name
+  * Image ID
+  * Port Mapping
+ * Information from ECS:
+  * Cluster Arn
+  * Container Instance Arn
+  * Task Arn
+
+ECS agent will create a file for each container under the directory `/var/lib/ecs/data/` and then mount the file to the container under a directory eg: `/ecs/metadata` with read only permission when the container is being created. Also an environment variable denoting the file path eg: `ECS_CONTAINER_METADATA_FILE` will be injected into the container. The `ECS_CONTAINER_METADATA_FILE` is consisted of several parts like: "/ecs/metadata/container_name_from_task_definition/metadata.json". After the container is started, ECS agent will update the metadata of container into this file. Then the container can get its metada information by parsing the json file with path `ECS_CONTAINER_METADATA_FILE`. This directory will be removed when task is being clean up. Also this feature will be configurable from the environment variable or in the task definition, eg: adding a field `"metadataEnable": true`. The detailed steps when start a container will be:
+* Check whether the feature is enabled, if not then go directly create and start the container.
+* Check the task definition make sure the path `/ecs` is not mounted, if it's used in the task definition, then disable the feature.
```

Comment:
If the `/ecs` path is in use, would it be possible to mount this metadata at a different point? The path is made available at `ECS_CONTAINER_METADATA_FILE` and therefore should not be statically addressed by a consumer.